### DISCUSSION TITLE
Add APT repository

### DISCUSSION THREAD

**Discussion by silopolis** [rustdesk/rustdesk#3985] [General]
Hi,

This great and so much needed piece of code deserves an APT repository to ease install and updates on Debian derivatives !

Thanks for all

TY
J

**Comment by simeononsecurity** [comment]
@silopolis I agree this would be nice, however if they did this, they would also have to implement repos for the other various OS'es. This would unfortunately require a fair bit of work to implement and eventually automate. However, I agree this should definitely happen at some point. However with the software in what I'd consider to be still beta, maybe even alpha, it's not the highest thing on the list.

**Reply by j-lakeman** [reply]
Related: https://github.com/rustdesk/rustdesk/discussions/7104

**Comment by Oceanwatcher** [comment]
I second this. A Ubuntu repository would first of all make it really easy to install. Just do a sudo apt install rustdesk.

But it would also make it real easy to keep updated as you can set the system to do minor updates automatically. This improves the security and should count for a lot. Actually, the upgrade process today is not documented as far as I can see. It should be as easy as sudo apt update and sudo apt upgrade ;-)

I avoid docker as the plague and prefer to have a small VPS for the server instead. And these days you can set up a small VPS on Oracle for free. It will at least help with the testing.

And I completely disagree that it has do be done for all other OS'es as well. At least not from the start.Ubuntu is a solid, widely used OS and would be a natural place to start. If other OS'es has similar delivery methods, then yes, it should be natural to add them eventually. But at least getting one OS up and running with this will lower the threshold for people to set up their own relay server.

**Comment by hex-m** [comment]
The most sustainable way to solve this would be to add rustdesk to the official repository of Debian. Ubuntu automatically gets packages that are in Debian ([src](https://wiki.ubuntu.com/UbuntuDevelopment/NewPackages)).

> Packages that have recently been added to Debian unstable will be automatically synced into Ubuntu

So the next step would be a [Debian RFP](https://wiki.debian.org/RFP).

**Comment by lweberk** [comment]
In the meantime I've been distributing it over a private repository, for which the problem becomes the versioning. Development versions (nightly) do not tag the dev portion (pre-release for ex.) in their release schema. It makes it really hard to propagate new versions this way, since there is nothing telling the packaging managers about the new nightly actually being newer than the one the day before.

I can imagine, this will also be a problem upstream.

**Reply by xlionjuan** [reply]
``rustdesk`` for stable, and ``rustdesk-next`` or ``rustdesk-nightly`` for nightly, for versioning problem, I think use *timestamp* should be okay. Or use git commit ID as version number.

**Comment by simeononsecurity** [comment]
It's easy, maintain a separate repo for nightly.
If you're going to install a nightly, potentially buggy release, it should be specific and determined by the users. 
The base install, that should be going to debian should be somewhat stable...

**Comment by Jackenmen** [comment]
> In the meantime I've been distributing it over a private repository, for which the problem becomes the versioning. Development versions (nightly) do not tag the dev portion (pre-release for ex.) in their release schema. It makes it really hard to propagate new versions this way, since there is nothing telling the packaging managers about the new nightly actually being newer than the one the day before.

I do this as well and I just reversion those deb files with a release counter, i.e. my current rustdesk version is `1.2.0-0gemfury64`. `deb-reversion` could be used, if not for the fact that rustdesk's deb file doesn't actually contain [the required `debian/changelog` file](https://www.debian.org/doc/manuals/maint-guide/dreq.en.html#changelog). If that file were there, it's simply a matter of running something like `fakeroot deb-reversion -v 1.2.0-0jackenmen1 rustdesk.deb`. Personally, I just decided to [reimplement what `deb-reversion` does in Python](https://github.com/Jackenmen/sync-debs-to-gemfury/blob/main/src/sync_debs_to_gemfury/deb_reversion.py) with some slight modifications and use that instead but asking for a `debian/changelog` file in another issue is also an option :)

**Comment by mofosyne** [comment]
Just wondering on the status of this effort as it would be nice to just run 

```
sudo apt-get install rustdesk
```

**Reply by xlionjuan** [reply]
I did it!
https://github.com/xlionjuan/rustdesk-apt-repo-latest

**Reply by jarbelix** [reply]
I already had rustdesk installed:

```
# dpkg -l | grep rustdesk 
ii rustdesk 1.3.0 amd64 A remote control software.

```
After following your tutorial I did an update quite easily.

```
# apt-get update
# apt-get dist-upgrade -y
```

Note that there was an update from 1.3.0 to 1.3.1

`# dpkg -l | grep rustdesk
ii rustdesk 1.3.1 amd64 A remote control software.
`

Congratulations on the excellent work.

**Comment by snuggle62a** [comment]
An alternative would be a snap or flatpak version.

**Reply by scop** [reply]
https://flathub.org/en/apps/com.rustdesk.RustDesk

Not sure why it is not mentioned in the release download docs.

**Reply by xlionjuan** [reply]
> https://flathub.org/en/apps/com.rustdesk.RustDesk
> 
> Not sure why it is not mentioned in the release download docs.

Because sandboxed version of RustDesk can't being unattended access.

**Comment by lweberk** [comment]
Does anyone know whether Amazon AWS, Cloudflare or any other provider grants resources for opensource stuff? To perhaps host our own downstream of this repository?

Want to clarify; I mean for the purpose of hosting a deb package repository that can actually be used as a sources list instead of manually downloading and installing the releases.

**Reply by xlionjuan** [reply]
For Cloudflare WARP, they host their own repository repo.

**Comment by bhack** [comment]
There is an official ITP at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1038942

**Reply by xlionjuan** [reply]
It is better that RustDesk host their own repositoriy, it could host on GitHub Pages and could publish by GitHub Actions

**Reply by Fantu** [reply]
Looking https://repology.org/project/rustdesk/versions very few distro packaging this program for now and seems only Arch from the major gnu/linux distro for now.
I don't see why it shouldn't have a package in the official Debian repository, the reasons why it couldn't (but I haven't checked and I don't know enough about Rustdesk) would be:
- copyright issue and/or use of licenses that don't follow [DFSG](https://wiki.debian.org/DebianFreeSoftwareGuidelines)
- rustdesk is don't is stable enough (excluding prerelease versions)
- it often makes changes that make it incompatible with older versions, requiring constant updating to the latest version
- it does not have good management of security issues, and reporting of any security fixes released

If rustdesk doesn't have one or more of the problems listed above I don't see the reason why it shouldn't be included in the official Debian repositories (if someone makes good packaging and maintains it)


From a fast look to deb (downloading from this repo latest release) the package content, excluding packaging things not optimal, I found strange the directory /usr/share/polkit-1 empty, it should contain a policy file normally.

**Reply by xlionjuan** [reply]
Debian tends to keep the packages old, but RustDesk is feature-rich and **rapid iteration** software, making users using old version for this kind of softwares are  ***not*** suggested.

For polkit problem: 

RustDesk is using polkit (`pkexec`) for unlock settings previously, but it cause some problems, so they use another way to achieve this, but polkit is still better for sure, because it supports fingerprint and other alternative authentication methods.

<hr>

I recommend to check out this video from Brodie Robertson
### Note: I'm not mean RustDesk hates repackaging, but for a perspective that why repackaging may cause problems
Title: Bottles Still Hates Linux Distro Packaging 
https://www.youtube.com/watch?v=WpuwIo2y4a4

**Reply by Fantu** [reply]
Having system packages that use system libraries (mostly linked) has its pros and cons, just as static builds or with upstream packages that include all or most dependencies inside (using for example appimage flatpak snap) have their own pros and cons.

From what I've seen, official Debian packages, if well maintained and if well maintained the dependencies also have an excellent quality, you could have even fewer problems, while if they are not maintained enough or there are some cases of software for which this is not convenient (I don't know if Rustdesk is one of those) then it may be better otherwise.
With software outside the packages from what I've seen the probability of bugs/unforeseen events is higher (then it depends on many factors), while more generally more space is consumed and some cases also have an additional weight and latency and possibility of unforeseen events given by the packaging itself used (the one where I had the most problems for example is snap).
Another problem with incorporating dependencies is that they are tied to the package itself, so you could have security bugs (and not only) that are normally fixed in the system packages, in case of serious bugs maybe even quickly, while you have external packages that might not fix them until the next version of that software or not even then (for example using versions of the fixed dependencies updated more rarely).
However it is difficult to explain everything well and it would take so much time that I have not.

There are also particular cases, like wine, even more complicated and for which it can easily be useful to have even different versions installed for different uses due to the ease of specific regressions (I think not due to poor quality of the software itself in that case but to the difficulty of developing something without enough specifications but very much for reverse engineering).

I hope that rustdesk grows and improve independently of the packaging because it seems very interesting to me, anyway I have tried it little for now.
For work (and also outside of it) I currently use a lot anydesk, unfortunately it is proprietary and I was thinking about its possible replacement with rustdesk in the future, when I will have enough time to test it well (make sure it works well for all the different devices we manage), prepare the selfhost part, and there will be no other more priority projects (currently there are many^^').
One thing I noticed with anydesk is that it seems to work even when connecting to devices with very old anydesk versions, I would be interested to know if it would be the same with rustdesk, not necessarily because of the desire not to update but because there are for example also cases of devices (or vm) with very old systems perhaps turned on very rarely (but in which assistance is needed in case of unforeseen events), an example is the consultation of old non-migratable management software (for accounting and not only unfortunately it is full of proprietary software that is difficult to transfer to other software or perhaps abandoned due to the bankruptcy of the companies that manage them)

**Reply by xlionjuan** [reply]
> One thing I noticed with anydesk is that it seems to work even when connecting to devices with very old anydesk versions, I would be interested to know if it would be the same with rustdesk

Yes, RustDesk often works when two devices' version gap are huge, but especially on Linux, it is better to keep RustDesk updated, because Linux Desktop has rapid iteration in recent years, I think I don't need to explain this.

**Comment by xlionjuan** [comment]
Unofficial repo:

https://github.com/xlionjuan/rustdesk-apt-repo-latest

**Comment by xlionjuan** [comment]
Fedora users, here we go with rpm repo!

https://github.com/xlionjuan/rustdesk-rpm-repo

**Comment by scop** [comment]
I think this is an essential feature to have. Considering the nature of the software, automatic updates are a must. RustDesk is often installed on systems whose users' skills are not up to updating the software manually nor do they have even interest in figuring out if updates are available.

Regarding providing repositories, I'm willing to chime in on making that happen on a repo hosting provider, as in helping out writing CI pipelines. I'll note a few free hosting services that I know of could be worth considering below.

https://packagecloud.io, plans: https://packagecloud.io/pricing/
The free plan might be a tad small for RustDesk. Storage use could be kept under control by pruning old releases.
I have first hand experience with this, has worked well for my (small) purposes. It's simple to use, requires a bit of setup in the service beforehand, but after that, CI setup is essentially as simple as https://github.com/aakso/ssh-inscribe/blob/b0d25e1ee25e32a6e8bf3fefcc642fdcf42deb4d/.github/workflows/release.yaml#L24-L32
Supports deb, rpm, etc, with different distro flavor repos, an a bunch of other package formats.

https://cloudsmith.com, plans: https://cloudsmith.com/pricing
Free plan is nominally even smaller than packagecloud.io, but there's a note at the end of the pricing/plans page:
> We love Open Source
> We offer a Free Open-Source plan with 50 GB artifact data and 200 GB package delivery. So long as your repository is public and has a valid Open-Source license, we’ve got you covered.
> Contact us for Details

I don't have first hand experience with this.
Supports a lot of package formats, IDK how it is for e.g. different deb/rpm distro flavors: https://cloudsmith.com/product/formats

**Reply by xlionjuan** [reply]
My approach didn't cost money 🤣

**Reply by scop** [reply]
Neither do the above. (Free plans have limits though.)

Since it was brought to the discussion, I would personally consider using a third party repo if the packages were signed by the origin project and my package manager would automatically verify them. I believe this is not the case nor really doable with deb/apt (could be with rpm where the actual packages are signed, but I don't use a rpm based distro, so not really interested in that myself).

**Reply by xlionjuan** [reply]
None of the Linux packages RustDesk ever signed, including `.rpm`, so my rpm repo will not only sign the repomd file, but also `.rpm` file.

Also enforce the repomd+package signing at provided `.repo` configuration file.

https://github.com/xlionjuan/rustdesk-rpm-repo/blob/main/wwwroot/latest.repo

**Reply by scop** [reply]
Understood, and I see those repos are trying to do the best they can. For the stated reasons that is not enough/interesting for me personally though, the possibilities for that just aren't there. Therefore proposing improvements to the RustDesk project proper and offering to help.

**Reply by xlionjuan** [reply]
On dpkg ecosystem, the package itself can't being verify via signing, they only relying on the package repository's metadata's signing 

On rpm world, both repomd(metadata) and `.rpm` can be signed, so we can see some weird things like the distros' repomd didn't signed, but `.rpm` did, or vise versa, or you can see multiple signature used on same repo, but as long as the gpg public key is installed and trusted, it won't break.

**Reply by scop** [reply]
Yes, that was noted in my grandparent message ~20 minutes ago. That is what makes a 3rd party repo solution that is interesting to me impossible.

**Reply by xlionjuan** [reply]
They decided not to, even the flathub submission is done by community initially 


B2y, you could download my shell script and do some slightly changes, and you'll get a solid script that auto download the latest release and install, so you won't need to rely on the repo by 3rd party.

**Reply by xlionjuan** [reply]
> Yes, that was noted in my grandparent message ~20 minutes ago. That is what makes a 3rd party repo solution that is interesting to me impossible.

GitHub Attestations and Sigstore Cosign exists, but they can't be integrated to package managers, unless you write a shell script...... they're meant for integrating to CI/CD, not for this kind of usecase.

**Reply by scop** [reply]
Since this offer hasn't caught traction, I'm withdrawing in the sense that I'll be unsubscribing from the issue. Feel free to ping me if there is interest towards this sometime and help is welcome. In the meantime flatpaks from flathub provide an alternative that is at least provided by this project and updates automatically (not sure about the integrity protections built in that).

**Comment by stdedos** [comment]
Host an APT repository on Github instructions:

https://gist.github.com/ikbelkirasan/57e60bbcaecd0bd5cf2cbf441e744159

**Reply by xlionjuan** [reply]
I've done that with GitHub Actions and Pages.

**Reply by stdedos** [reply]
I know. ... but you'll understand if I don't want to trust your key :pray: 

I am just trying to nudge ... https://github.com/rustdesk/rustdesk

**Comment by arnowelzel** [comment]
Setting up a repository which can be used with APT is really not that hard. You can also integrate this to Github actions, so the repository will automatically be updated with every new release. It would be extemely helpful if Rustdesk would provide this. Is any help needed for this?

**Reply by xlionjuan** [reply]
I'm already using the repos that I created for years, but they still decided to not take over :)

**Reply by jarbelix** [reply]
I've been using your repository for quite some time on my Raspberry Pis, which are located in some bus stations in the state of Mato Grosso do Sul, Brazil.

Your work is fantastic. Congratulations.

https://wiki.tiozaodolinux.com/Guide-for-Linux/Zabbix-RaspBerry-PI#install-rustdesk

**Reply by arnowelzel** [reply]
I think you refer to https://github.com/xlionjuan/rustdesk-apt-repo-latest

Do you accept donations? I'd like to support your work.

**Comment by Dsw77** [comment]
I created a script to get the latest .deb version from the official rustdesk git repo, combined with a timer that runs the script every day.
https://github.com/Dsw77/rustdesk-updater-debian
